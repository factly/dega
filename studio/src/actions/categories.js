import axios from 'axios';
import {
  ADD_CATEGORIES,
  ADD_CATEGORIES_REQUEST,
  SET_CATEGORIES_LOADING,
  RESET_CATEGORIES,
  CATEGORIES_API,
  GET_CATEGORY,
  UPDATE_CATEGORY,
} from '../constants/categories';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addMedia } from './media';
import getError from '../utils/getError';

// action to fetch all categories
export const getCategories = (query) => {
  return (dispatch, getState) => {
    const currentSpaceID = getState().spaces?.selected;
    if (currentSpaceID === 0) {
      return;
    }
    dispatch(loadingCategories());
    return axios
      .get(CATEGORIES_API, {
        params: query,
      })
      .then((response) => {
        dispatch(
          addMedia(
            response.data.nodes
              .filter((category) => category.medium)
              .map((category) => category.medium),
          ),
        );
        dispatch(
          addCategoriesList(
            response.data.nodes.map((category) => {
              return { ...category, medium: category.medium?.id };
            }),
          ),
        );
        dispatch(
          addCategoriesRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to fetch category by id
export const getCategory = (id) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .get(CATEGORIES_API + '/' + id)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));

        dispatch(addCategory(GET_CATEGORY, { ...response.data, medium: response.data.medium?.id }));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to create category
export const createCategory = (data) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .post(CATEGORIES_API, data)
      .then(() => {
        dispatch(resetCategories());
        dispatch(addSuccessNotification('Category created'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update category by id
export const updateCategory = (data) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .put(CATEGORIES_API + '/' + data.id, data)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));

        dispatch(
          addCategory(UPDATE_CATEGORY, { ...response.data, medium: response.data.medium?.id }),
        );
        dispatch(addSuccessNotification('Category updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to delete category by id
export const deleteCategory = (id) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .delete(CATEGORIES_API + '/' + id)
      .then(() => {
        dispatch(resetCategories());
        dispatch(addSuccessNotification('Category deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const addCategories = (categories) => {
  return (dispatch) => {
    dispatch(
      addMedia(categories.filter((category) => category.medium).map((category) => category.medium)),
    );
    dispatch(
      addCategoriesList(
        categories.map((category) => {
          return { ...category, medium: category.medium?.id };
        }),
      ),
    );
  };
};

export const loadingCategories = () => ({
  type: SET_CATEGORIES_LOADING,
  payload: true,
});

export const stopCategoriesLoading = () => ({
  type: SET_CATEGORIES_LOADING,
  payload: false,
});

export const addCategory = (type, payload) => ({
  type,
  payload,
});

export const addCategoriesList = (data) => ({
  type: ADD_CATEGORIES,
  payload: data,
});

export const addCategoriesRequest = (data) => ({
  type: ADD_CATEGORIES_REQUEST,
  payload: data,
});

export const resetCategories = () => ({
  type: RESET_CATEGORIES,
});
