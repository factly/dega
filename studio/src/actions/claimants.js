import axios from 'axios';
import {
  ADD_CLAIMANTS,
  ADD_CLAIMANTS_REQUEST,
  SET_CLAIMANTS_LOADING,
  RESET_CLAIMANTS,
  CLAIMANTS_API,
  UPDATE_CLAIMANT,
  GET_CLAIMANT,
} from '../constants/claimants';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addMedia } from './media';
import getError from '../utils/getError';

// action to fetch all claimants
export const getClaimants = (query) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .get(CLAIMANTS_API, {
        params: query,
      })
      .then((response) => {
        dispatch(
          addMedia(
            response.data.nodes
              .filter((claimant) => claimant.medium)
              .map((claimant) => claimant.medium),
          ),
        );
        dispatch(
          addClaimantsList(
            response.data.nodes.map((claimant) => {
              claimant.description = {
                json: claimant.description,
                html: claimant.html_description,
              };
              return { ...claimant, medium: claimant.medium?.id };
            }),
          ),
        );
        dispatch(
          addClaimantsRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimantsLoading()));
  };
};

// action to fetch claimant by id
export const getClaimant = (id) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .get(CLAIMANTS_API + '/' + id)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));
        response.data.description = {
          json: response.data.description,
          html: response.data.html_description,
        };
        dispatch(addClaimant(GET_CLAIMANT, { ...response.data, medium: response.data.medium?.id }));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimantsLoading()));
  };
};

// action to create claimant
export const createClaimant = (data) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .post(CLAIMANTS_API, data)
      .then(() => {
        dispatch(resetClaimants());
        dispatch(addSuccessNotification('Claimant created'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update claimant by id
export const updateClaimant = (data) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .put(CLAIMANTS_API + '/' + data.id, data)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));
        response.data.description = {
          json: response.data.description,
          html: response.data.html_description,
        };
        dispatch(
          addClaimant(UPDATE_CLAIMANT, { ...response.data, medium: response.data.medium?.id }),
        );
        dispatch(addSuccessNotification('Claimant updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimantsLoading()));
  };
};

// action to delete claimant by id
export const deleteClaimant = (id) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .delete(CLAIMANTS_API + '/' + id)
      .then(() => {
        dispatch(resetClaimants());
        dispatch(addSuccessNotification('Claimant deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const addClaimants = (claimants) => {
  return (dispatch) => {
    dispatch(
      addMedia(claimants.filter((claimant) => claimant.medium).map((claimant) => claimant.medium)),
    );
    dispatch(
      addClaimantsList(
        claimants.map((claimant) => {
          return { ...claimant, medium: claimant.medium?.id };
        }),
      ),
    );
  };
};

export const loadingClaimants = () => ({
  type: SET_CLAIMANTS_LOADING,
  payload: true,
});

export const stopClaimantsLoading = () => ({
  type: SET_CLAIMANTS_LOADING,
  payload: false,
});

export const addClaimant = (type, payload) => ({
  type,
  payload,
});

export const addClaimantsList = (data) => ({
  type: ADD_CLAIMANTS,
  payload: data,
});

export const addClaimantsRequest = (data) => ({
  type: ADD_CLAIMANTS_REQUEST,
  payload: data,
});

export const resetClaimants = () => ({
  type: RESET_CLAIMANTS,
});
