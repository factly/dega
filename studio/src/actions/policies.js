import axios from 'axios';
import {
  ADD_POLICIES,
  ADD_POLICIES_REQUEST,
  SET_POLICIES_LOADING,
  RESET_POLICIES,
  POLICIES_API,
  UPDATE_POLICY,
  GET_POLICY,
} from '../constants/policies';
import { addErrorNotification, addSuccessNotification } from './notifications';
import getError from '../utils/getError';

// action to fetch all policies
export const addDefaultPolicies = (query) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .post(POLICIES_API + '/default')
      .then((response) => {
        dispatch(addPolicies(response.data.nodes));
        dispatch(
          addPoliciesRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopPoliciesLoading()));
  };
};

// action to fetch all policies
export const getPolicies = (query) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .get(POLICIES_API, {
        params: query,
      })
      .then((response) => {
        dispatch(addPolicies(response.data.nodes));
        dispatch(
          addPoliciesRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopPoliciesLoading()));
  };
};

// action to fetch policy by id
export const getPolicy = (id) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .get(POLICIES_API + '/' + id)
      .then((response) => {
        dispatch(addPolicy(GET_POLICY, response.data));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopPoliciesLoading()));
  };
};

// action to create policy
export const createPolicy = (data) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .post(POLICIES_API, data)
      .then(() => {
        dispatch(resetPolicies());
        dispatch(addSuccessNotification('Policy created'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update policy by id
export const updatePolicy = (data) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .put(POLICIES_API + '/' + data.id, data)
      .then((response) => {
        dispatch(addPolicy(UPDATE_POLICY, response.data));
        dispatch(addSuccessNotification('Policy updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopPoliciesLoading()));
  };
};

// action to delete policy by id
export const deletePolicy = (id) => {
  return (dispatch) => {
    dispatch(loadingPolicies());
    return axios
      .delete(POLICIES_API + '/' + id)
      .then(() => {
        dispatch(resetPolicies());
        dispatch(addSuccessNotification('Policy deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const loadingPolicies = () => ({
  type: SET_POLICIES_LOADING,
  payload: true,
});

export const stopPoliciesLoading = () => ({
  type: SET_POLICIES_LOADING,
  payload: false,
});

export const addPolicy = (type, payload) => ({
  type,
  payload,
});

export const addPolicies = (payload) => ({
  type: ADD_POLICIES,
  payload,
});

export const addPoliciesRequest = (payload) => ({
  type: ADD_POLICIES_REQUEST,
  payload,
});

export const resetPolicies = () => ({
  type: RESET_POLICIES,
});
