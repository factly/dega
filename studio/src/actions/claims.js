import axios from 'axios';
import {
  ADD_CLAIMS,
  ADD_CLAIMS_REQUEST,
  SET_CLAIMS_LOADING,
  RESET_CLAIMS,
  CLAIMS_API,
  GET_CLAIM,
  UPDATE_CLAIM,
} from '../constants/claims';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addRatings } from './ratings';
import { addClaimants } from './claimants';
import getError from '../utils/getError';

// action to fetch all claims
export const getClaims = (query) => {
  const params = new URLSearchParams();
  if (query.claimant && query.claimant.length > 0) {
    query.claimant.map((each) => params.append('claimant', each));
  }
  if (query.rating && query.rating.length > 0) {
    query.rating.map((each) => params.append('rating', each));
  }
  if (query.page) {
    params.append('page', query.page);
  }
  if (query.limit) {
    params.append('limit', query.limit);
  }
  if (query.sort) {
    params.append('sort', query.sort);
  }
  if (query.q) {
    params.append('q', query.q);
  }
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .get(CLAIMS_API, {
        params: params,
      })
      .then((response) => {
        dispatch(
          addClaimants(
            response.data.nodes.filter((claim) => claim.claimant).map((claim) => claim.claimant),
          ),
        );
        dispatch(
          addRatings(
            response.data.nodes.filter((claim) => claim.rating).map((claim) => claim.rating),
          ),
        );
        dispatch(
          addClaimsList(
            response.data.nodes.map((claim) => {
              claim.description = { json: claim.description, html: claim.description_html };
              return { ...claim, claimant: claim.claimant.id, rating: claim.rating.id };
            }),
          ),
        );
        dispatch(
          addClaimsRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimsLoading()));
  };
};

// action to fetch claim by id
export const getClaim = (id) => {
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .get(CLAIMS_API + '/' + id)
      .then((response) => {
        let claim = response.data;
        claim.description = { json: claim.description, html: claim.description_html };
        dispatch(addClaimants([claim.claimant]));
        dispatch(addRatings([claim.rating]));

        dispatch(
          addClaim(GET_CLAIM, { ...claim, claimant: claim.claimant.id, rating: claim.rating.id }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimsLoading()));
  };
};

// action to create claim
export const createClaim = (data) => {
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .post(CLAIMS_API, data)
      .then((response) => {
        let claim = response.data;
        claim.description = { json: claim.description, html: claim.description_html };
        dispatch(addClaimants([claim.claimant]));
        dispatch(addRatings([claim.rating]));

        dispatch(resetClaims());
        dispatch(addSuccessNotification('Claim created'));
        return claim;
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update claim by id
export const updateClaim = (data) => {
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .put(CLAIMS_API + '/' + data.id, data)
      .then((response) => {
        let claim = response.data;
        claim.description = { json: claim.description, html: claim.description_html };
        dispatch(addClaimants([claim.claimant]));
        dispatch(addRatings([claim.rating]));

        dispatch(
          addClaim(UPDATE_CLAIM, {
            ...claim,
            claimant: claim.claimant.id,
            rating: claim.rating.id,
          }),
        );
        dispatch(addSuccessNotification('Claim updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimsLoading()));
  };
};

// action to delete claim by id
export const deleteClaim = (id) => {
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .delete(CLAIMS_API + '/' + id)
      .then(() => {
        dispatch(resetClaims());
        dispatch(addSuccessNotification('Claim deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const addClaims = (claims) => {
  return (dispatch) => {
    dispatch(addClaimants(claims.filter((claim) => claim.claimant).map((claim) => claim.claimant)));
    dispatch(addRatings(claims.filter((claim) => claim.rating).map((claim) => claim.rating)));
    dispatch(
      addClaimsList(
        claims.map((claim) => {
          return { ...claim, claimant: claim.claimant.id, rating: claim.rating.id };
        }),
      ),
    );
  };
};

export const loadingClaims = () => ({
  type: SET_CLAIMS_LOADING,
  payload: true,
});

export const stopClaimsLoading = () => ({
  type: SET_CLAIMS_LOADING,
  payload: false,
});

export const addClaim = (type, payload) => ({
  type,
  payload,
});

export const addClaimsList = (payload) => ({
  type: ADD_CLAIMS,
  payload,
});

export const addClaimsRequest = (payload) => ({
  type: ADD_CLAIMS_REQUEST,
  payload,
});

export const resetClaims = () => ({
  type: RESET_CLAIMS,
});
