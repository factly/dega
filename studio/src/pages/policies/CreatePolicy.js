import React from 'react';
import { createPolicy } from '../../actions/policies';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import PolicyCreateForm from './components/PolicyForm';
import { Helmet } from 'react-helmet';

function CreatePolicy() {
  const history = useHistory();

  const dispatch = useDispatch();
  const onCreate = (values) => {
    values.permissions = values.permissions?.filter(
      (item) => item && item.resource && item.actions.length > 0,
    );
    dispatch(createPolicy(values)).then(() => history.push('/members/policies'));
  };

  return (
    <>
      <Helmet title={'Create Policy'} />
      <PolicyCreateForm onCreate={onCreate} />
    </>
  );
}

export default CreatePolicy;
