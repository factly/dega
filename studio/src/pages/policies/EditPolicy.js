import React from 'react';
import PolicyEditForm from './components/PolicyForm';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton } from 'antd';
import { getPolicy, updatePolicy } from '../../actions/policies';
import { useHistory } from 'react-router-dom';
import { useParams } from 'react-router-dom';
import RecordNotFound from '../../components/ErrorsAndImage/RecordNotFound';
import { Helmet } from 'react-helmet';

function EditPolicy() {
  const history = useHistory();
  const { id } = useParams();

  const dispatch = useDispatch();

  const { policy, loading } = useSelector((state) => {
    if (!state.policies.details[id])
      return {
        policy: null,
        loading: state.policies.loading,
      };

    return {
      policy: {
        ...state.policies.details[id],
        permissions: state.policies.details[id].permissions.reduce(
          (obj, item) => Object.assign(obj, { [item.resource]: item.actions }),
          {},
        ),
        roles: state.policies.details[id].roles?.length 
          ? state.policies.details[id].roles.map((item) => item.id)
          : [],
      },
      loading: state.policies.loading,
    };
  });

  React.useEffect(() => {
    dispatch(getPolicy(id));
  }, [dispatch, id]);

  if (loading) return <Skeleton />;

  if (!policy) {
    return <RecordNotFound />;
  }

  const onUpdate = (values) => {
    dispatch(updatePolicy({ ...policy, ...values })).then(() =>
      history.push(`/members/policies/`),
    );
  };

  return (
    <>
      <Helmet title={`${policy.name} - Edit Policy`} />
      <PolicyEditForm data={policy} onCreate={onUpdate} />
    </>
  );
}

export default EditPolicy;
