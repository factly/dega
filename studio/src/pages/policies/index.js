import React from 'react';
import { Space, Button, Row } from 'antd';
import { Link, useLocation } from 'react-router-dom';
import PolicyList from './components/PolicyList';
import getUserPermission from '../../utils/getUserPermission';
import { useDispatch, useSelector } from 'react-redux';
import { getPolicies } from '../../actions/policies';
import deepEqual from 'deep-equal';
import Loader from '../../components/Loader';
import { Helmet } from 'react-helmet';

function Policies() {
  const spaces = useSelector(({ spaces }) => spaces);
  const actions = getUserPermission({ resource: 'policies', action: 'get', spaces });
  const dispatch = useDispatch();
  const query = new URLSearchParams(useLocation().search);
  const [filters, setFilters] = React.useState({
    page: 1,
    limit: 20,
  });
  query.set('page', filters.page);
  window.history.replaceState({}, '', `${window.PUBLIC_URL}${useLocation().pathname}?${query}`);
  const { policies, total, loading } = useSelector((state) => {
    const node = state.policies.req.find((item) => {
      return deepEqual(item.query, filters);
    });

    if (node)
      return {
        policies: node.data.map((element) => state.policies.details[element]),
        total: node.total,
        loading: state.policies.loading,
      };
    return { policies: [], total: 0, loading: state.policies.loading };
  });

  React.useEffect(() => {
    fetchPolicies();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [filters]);

  const fetchPolicies = () => {
    dispatch(getPolicies(filters));
  };

  return loading ? (
    <Loader />
  ) : (
    <Space direction="vertical">
      <Helmet title={'Policies'} />
      <Row gutter={16} justify="end">
        <Link to="/members/policies/create">
          <Button
            disabled={!(actions.includes('admin') || actions.includes('create'))}
            type="primary"
          >
            New Policy
          </Button>
        </Link>
      </Row>
      <PolicyList
        actions={actions}
        data={{ policies, total, loading }}
        filters={filters}
        setFilters={setFilters}
        fetchPolicies={fetchPolicies}
      />
    </Space>
  );
}

export default Policies;
