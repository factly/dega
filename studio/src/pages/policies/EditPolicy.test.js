import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import { act } from '@testing-library/react';

import '../../matchMedia.mock';
import EditPolicy from './EditPolicy';
import * as actions from '../../actions/policies';
import PolicyEditForm from './components/PolicyForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('react-monaco-editor', () => {
  const MonacoEditor = () => <div />;
  return MonacoEditor;
});

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
  useParams: jest.fn().mockReturnValue({ id: '1' }),
}));

jest.mock('../../actions/policies', () => ({
  getPolicies: jest.fn(),
  addPolicy: jest.fn(),
  getPolicy: jest.fn(),
  updatePolicy: jest.fn(),
}));

describe('Policies edit component', () => {
  let store;
  let mockedDispatch;
  store = mockStore({
    policies: {
      req: [],
      details: {
        '1': {
          id: 1,
          name: 'policy',
          description: 'description',
          permissions: [
            {
              resource: 'policy',
              actions: ['create'],
            },
          ],
        },
      },
      loading: false,
    },
    authors: {
      req: [{ query: { page: 1 }, total: 1, data: [1] }],
      details: { 1: { id: 1, name: 'Author', email: 'author@aut.co' } },
      loading: false,
    },
  });
  store.dispatch = jest.fn(() => ({}));
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);

  describe('snapshot testing', () => {
    it('should render the component', () => {
      let tree;
      act(() => {
        tree = shallow(
          <Provider store={store}>
            <EditPolicy />
          </Provider>,
        );
      });
      expect(tree).toMatchSnapshot();
    });
    it('should match component with empty data', () => {
      let tree;
      store = mockStore({
        policies: {
          req: [],
          details: {},
          loading: false,
        },
        authors: {
          req: [],
          details: {},
          loading: false,
        },
      });
      act(() => {
        tree = shallow(
          <Provider store={store}>
            <EditPolicy />
          </Provider>,
        );
      });
      expect(tree).toMatchSnapshot();
    });
    it('should match skeleton while loading', () => {
      let tree;
      store = mockStore({
        policies: {
          req: [],
          details: {},
          loading: true,
        },
        authors: {
          req: [],
          details: {},
          loading: false,
        },
      });
      act(() => {
        tree = shallow(
          <Provider store={store}>
            <EditPolicy />
          </Provider>,
        );
      });
      expect(tree).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    beforeEach(() => {
      store = mockStore({
        policies: {
          req: [],
          details: {
            '1': {
              id: 1,
              name: 'policy',
              description: 'description',
              permissions: [
                {
                  resource: 'policy',
                  actions: ['create'],
                },
              ],
            },
          },
          loading: false,
        },
        authors: {
          req: [{ query: { page: 1 }, total: 1, data: [1] }],
          details: { 1: { id: 1, name: 'Author', email: 'author@aut.co' } },
          loading: false,
        },
      });
    });
    afterEach(() => {
      wrapper.unmount();
    });
    it('should display RecordNotFound when no Policy found', () => {
      store = mockStore({
        policies: {
          req: [],
          details: {},
          loading: false,
        },
        authors: {
          req: [{ query: { page: 1 }, total: 1, data: [1] }],
          details: { 1: { id: 1, name: 'Author', email: 'author@aut.co' } },
          loading: false,
        },
      });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <EditPolicy />
          </Provider>,
        );
      });
      expect(wrapper.find('RecordNotFound').length).toBe(1);
    });
    it('should display Skeleton when loading', () => {
      store = mockStore({
        policies: {
          req: [],
          details: {},
          loading: true,
        },
        authors: {
          req: [{ query: { page: 1 }, total: 1, data: [1] }],
          details: { 1: { id: 1, name: 'Author', email: 'author@aut.co' } },
          loading: false,
        },
      });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <EditPolicy />
          </Provider>,
        );
      });
      expect(wrapper.find('Skeleton').length).toBe(1);
    });
    it('should call getAction and display policies with non empty users', () => {
      store = mockStore({
        policies: {
          req: [],
          details: {
            '1': {
              id: 1,
              name: 'policy',
              description: 'description',
              permissions: [
                {
                  resource: 'policy',
                  actions: ['create'],
                },
              ],
              users: [{ id: 10, email: 'abc@gmail.com' }],
            },
          },

          loading: false,
        },
        authors: {
          req: [{ query: { page: 1 }, total: 1, data: [1] }],
          details: { 1: { id: 1, name: 'Author', email: 'author@aut.co' } },
          loading: false,
        },
      });
      actions.getPolicy.mockReset();
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <EditPolicy />
          </Provider>,
        );
      });
      expect(actions.getPolicy).toHaveBeenCalledWith('1');
    });
    it('should call get action', () => {
      actions.getPolicy.mockReset();
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <EditPolicy />
          </Provider>,
        );
      });
      expect(actions.getPolicy).toHaveBeenCalledWith('1');
    });
    it('should call updatePolicy', (done) => {
      actions.updatePolicy.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      act(() => {
        wrapper = mount(
          <Provider store={store}>
            <EditPolicy />
          </Provider>,
        );
      });
      wrapper.find(PolicyEditForm).props().onCreate({ test: 'test' });
      setTimeout(() => {
        expect(actions.updatePolicy).toHaveBeenCalledWith({
          id: 1,
          name: 'policy',
          description: 'description',
          permissions: {
            policy: ['create'],
          },
          users: [],
          test: 'test',
        });
        expect(push).toHaveBeenCalledWith('/members/policies/1/edit');
        done();
      }, 0);
    });
  });
});
