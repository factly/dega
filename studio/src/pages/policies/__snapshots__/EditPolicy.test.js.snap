// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policies edit component snapshot testing should match component with empty data 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <EditPolicy />
</ContextProvider>
`;

exports[`Policies edit component snapshot testing should match skeleton while loading 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <EditPolicy />
</ContextProvider>
`;

exports[`Policies edit component snapshot testing should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <EditPolicy />
</ContextProvider>
`;
